### PR TITLE
Optimize W matrix formation and fix `fastconvergence` judgment

### PR DESCRIPTION
**Pull request by YingboMa**
`ηold` is approximately `ndz/ndzprev`, so the lower it is, the faster convergence the Newton iteration has.

### CODE REVIEW

**Comment by YingboMa** [comment]
I also changed `W = M - gamma*J` to `W = -M + gamma*J`

**Comment by YingboMa** [comment]
hires
rober
pollution

**Comment by YingboMa** [comment]
BNGNetwork:
```julia
julia> sol_cvodebdf = @time solve(prob, CVODE_BDF());
 29.039184 seconds (1.80 M allocations: 112.732 MiB, 0.83% gc time)

julia> sol_cvodebdf.destats
DiffEqBase.DEStats
Number of function 1 evaluations: 1309
Number of function 2 evaluations: -1
Number of W matrix evaluations: 152
Number of linear solves:        -1
Number of Jacobians created:        26
Number of nonlinear solver iterations:        1305
Number of nonlinear solver convergence failures:        0
Number of rootfind condition calls:        0
Number of accepted steps:       736
Number of rejected steps:       55

julia> sol_kencarp4 = @time solve(prob, KenCarp4(autodiff=false));
 24.699485 seconds (4.23 k allocations: 26.708 MiB, 0.03% gc time)

julia> sol_kencarp4.destats
DiffEqBase.DEStats
Number of function 1 evaluations: 275678
Number of function 2 evaluations: 0
Number of W matrix evaluations: 242
Number of linear solves:        4103
Number of Jacobians created:        242
Number of nonlinear solver iterations:        3909
Number of nonlinear solver convergence failures:        51
Number of rootfind condition calls:        0
Number of accepted steps:       193
Number of rejected steps:       1
```

File: src/derivative_utils.jl [hunk 1]
```diff
@@ -261,23 +263,29 @@ end
...
   @inbounds if W_transform
-    invdtgamma′ = inv(dtgamma)
-    for i in iijj[1]
-      @inbounds for j in iijj[2]
-        W[i, j] = muladd(mass_matrix[i, j], invdtgamma′, -J[i, j])
+    invdtgamma = inv(dtgamma)
+    if MT <: UniformScaling
+      copyto!(W, J)
+      @simd for i in diagind(W)
+          W[i] = muladd(-mass_matrix.λ, invdtgamma, J[i])
```
Comment (ChrisRackauckas) [inline src/derivative_utils.jl]: what if the diagonal term is zero in a sparse matrix? That could happen

Comment (YingboMa) [reply src/derivative_utils.jl]: This branch is only for dense matrices.

Comment (ChrisRackauckas) [reply src/derivative_utils.jl]: how so?

**Comment by YingboMa** [comment]
# Rober release
rober_release
# Rober PR
rober_PR


# Hires release
hires_release
# Hires PR
hires_PR

# Pollution release
pollution

# Pollution PR
pollution_PR

**Pull request merged**